### ISSUE TITLE
Support for async functions

### ISSUE THREAD

**Issue by martell** [google/evcxr#90]
```
:dep reqwest = "0.10"

let body = reqwest::get("https://www.rust-lang.org")
    .await.unwrap()
    .text()
    .await.unwrap();

println!("body = {:?}", body);
```

Errors:

```
let body = reqwest::get("https://www.rust-lang.org")
    .await.unwrap()
    .text()
    .await.unwrap();
            only allowed inside `async` functions and blocks
`await` is only allowed inside `async` functions and blocks

let body = reqwest::get("https://www.rust-lang.org")
    .await.unwrap()
            only allowed inside `async` functions and blocks
`await` is only allowed inside `async` functions and blocks
```

**Comment by davidlattimore** [comment]
Nice suggestion. I guess an executor of some sort would be needed. Any idea for how you'd expect that to work? I'm guessing a really simple executor that just called poll repeatedly until it returns Ready would suffice?

**Comment by martell** [comment]
> I'm guessing a really simple executor that just called poll repeatedly until it returns Ready would suffice?

Seems look a good idea to me, I'm guessing looking at `basic_scheduler` in tokio as a reference would be a good choice?

Do we have a way to detect based on the content (keywords ? that might now be great) that the user expects an async function?

I've only just really started to dig into the async world.
I know that `async` keyword wraps the return types in a future but I doubt it does some magic for the actual value being returned ? Having to use `async move` to return something where it did not need to be an async function would be not great?

**Comment by thomaseizinger** [comment]
> Nice suggestion. I guess an executor of some sort would be needed. Any idea for how you'd expect that to work? I'm guessing a really simple executor that just called poll repeatedly until it returns Ready would suffice?

Would there be any downside to just spawning a tokio runtime inside the temporary project and making the function async?

**Comment by davidlattimore** [comment]
I'm somewhat inclined to agree. I guess the downside would be that I gather it's a relatively heavy dependency, so would probably slow down linking a bit.

**Comment by davidB** [comment]
https://crates.io/crates/async-std should be lighter than tokio (I'm a noob
in rust async)

**Comment by thomaseizinger** [comment]
> [crates.io/crates/async-std](https://crates.io/crates/async-std) (task) should be lighter for the usage of evcxr than tokio (I'm a noob in rust async)

Tokio 0.2 also has a task system :)
Without actual benchmarks, it is hard to say which one is going to be heavier.

**Comment by davidlattimore** [comment]
I just had a look at Tokio. Turned out to be relatively simple.

The following works:

```rust
:dep reqwest = "0.10"
:dep tokio = "0.2"
tokio::runtime::Runtime::new().unwrap().block_on(async {
    let body = reqwest::get("http://localhost/2")
        .await.unwrap()
        .text()
        .await.unwrap();

    println!("body = {:?}", body);
})
```

If I then rerun that with minor edits (and without repeating the :dep commands), it takes about 1000ms on my machine (opt-level 0 with lld as the linker). But it seems that most of the slowdown is actually from Reqwest, not from Tokio.

In a separate notebook, with only Tokio loaded, the following takes a little under 400ms after minor edits:

```rust
tokio::runtime::Runtime::new().unwrap().block_on(async {
    println!("Hello, World 4!");
})
```

I think it should be relatively straightforward to detect the presence of the await keyword (evcxr already parses the source code) and add tokio as a dependency then wrap the user code with something like what I did above. Perhaps we should also put the behaviour behind an option such as `:async tokio` (could still be the default), but that would allow people to turn if off, if they wanted to set up the runtime themselves as I did above.
